dashboard/modules/serve/tests/test_serve_dashboard.py

```diff
@@ -426,6 +431,103 @@ def applications_running():
         assert app_details[app].last_deployed_time_s > 0
         assert app_details[app].route_prefix == expected_values[app]["route_prefix"]
         assert app_details[app].docs_path == expected_values[app]["docs_path"]
+        assert app_details[app].source == expected_values[app]["source"]
+
+        # CHECK: all deployments are present
+        assert (
+            app_details[app].deployments.keys() == expected_values[app]["deployments"]
+        )
+
+        for deployment in app_details[app].deployments.values():
+            assert deployment.status == DeploymentStatus.HEALTHY
+            assert (
+                deployment.status_trigger
+                == DeploymentStatusTrigger.CONFIG_UPDATE_COMPLETED
+            )
+            # Route prefix should be app level options eventually
+            assert "route_prefix" not in deployment.deployment_config.dict(
+                exclude_unset=True
+            )
+            if isinstance(deployment.deployment_config.num_replicas, int):
+                assert (
+                    len(deployment.replicas)
+                    == deployment.deployment_config.num_replicas
+                )
+                assert len(deployment.replicas) == deployment.target_num_replicas
+
+            for replica in deployment.replicas:
+                assert replica.replica_id
+                assert replica.state == ReplicaState.RUNNING
+                assert deployment.name in replica.actor_name
+                assert replica.actor_id and replica.node_id and replica.node_ip
+                assert replica.start_time_s > app_details[app].last_deployed_time_s
+                file_path = "/tmp/ray/session_latest/logs" + replica.log_file_path
+                assert os.path.exists(file_path)
+
+    print("Finished checking application details.")
+
+
+@pytest.fixture
+def serve_start_and_shutdown(ray_start_stop: None) -> Iterator[None]:
+    """
+    See https://github.com/ray-project/ray/pull/45522#issuecomment-2159403223
+    """
+    serve.start()
+    yield
+    serve.shutdown()
+
+
+@pytest.mark.skipif(
+    sys.platform == "darwin" and not TEST_ON_DARWIN, reason="Flaky on OSX."
+)
+@pytest.mark.parametrize("url", [SERVE_AGENT_URL, SERVE_HEAD_URL])
+def test_get_serve_instance_details_for_imperative_apps(serve_start_and_shutdown, url):
+    """
+    Most behavior is checked by test_get_serve_instance_details.
+    This test mostly checks for the different behavior of
+    imperatively-deployed apps, with some crossover.
+    """
+    serve.run(world.DagNode, name="app1", route_prefix="/apple")
+    serve.run(world.DagNode, name="app2", route_prefix="/banana")
```

Comment:
@JoshKarpel Since it's likely these two lines that's causing the problems, can you try submitting these as a separate subprocess?

Reply:
Sorry for the delay, but this seems to have worked! Thanks for the suggestion. I'm gonna do a quick self-review and then put this back up for review.